File: sort_examples/quickSort.js
sort = (array, film) => {
  let compares = 0;

  const partition = (left, right) => {
    const pivot = array[right];
    let i = left;
    for (let j = left; j < right; j++) {
      film.rec({ array, i, j, temp: pivot, compares });
      compares++;
      if (array[j] < pivot) {
        const temp = array[i];
        array[i] = array[j];
        array[j] = temp;
        film.rec({ array, i, j, temp: pivot, compares });
        i++;
      }
    }
    const temp = array[i];
    array[i] = array[right];
    array[right] = temp;
    film.rec({ array, i, j: right, temp: pivot, compares });
    return i;
  };

  const quickSort = (left, right) => {
    if (left >= right) return;
    const pivotIndex = partition(left, right);
    quickSort(left, pivotIndex - 1);
    quickSort(pivotIndex + 1, right);
  };

  quickSort(0, array.length - 1);
};

File: sort_examples/quick_sort.py
def sort(array):
    def partition(left, right):
        pivot = array[right]
        array.note("pivot", pivot)
        destination = left

        for index in range(left, right):
            if array[index] < pivot:
                array.swap(destination, index)
                destination += 1

        array.swap(destination, right)
        array.mark_sorted(destination)
        return destination

    def quick_sort(left, right):
        if left > right:
            return
        if left == right:
            array.mark_sorted(left)
            return

        pivot_index = partition(left, right)
        quick_sort(left, pivot_index - 1)
        quick_sort(pivot_index + 1, right)

    quick_sort(0, len(array) - 1)
